src/combinations.rs

```diff
@@ -27,22 +27,70 @@ impl<I> fmt::Debug for Combinations<I>
     debug_fmt_fields!(Combinations, indices, pool, first);
 }
 
-/// Create a new `Combinations` from a clonable iterator.
+/// Create a new `Combinations` from a cloneable iterator.
 pub fn combinations<I>(iter: I, k: usize) -> Combinations<I>
     where I: Iterator
 {
-    let mut pool: LazyBuffer<I> = LazyBuffer::new(iter);
+    let mut pool = LazyBuffer::new(iter);
+    pool.prefill(k);
 
-    for _ in 0..k {
-        if !pool.get_next() {
-            break;
+    Combinations::from_pool(pool, k)
+}
+
+impl<I: Iterator> Combinations<I> {
+    #[inline]
+    pub(crate) fn from_pool(pool: LazyBuffer<I>, k: usize) -> Self {
+        Combinations {
+            indices: (0..k).collect(),
+            pool,
+            first: true,
         }
     }
 
-    Combinations {
-        indices: (0..k).collect(),
-        pool,
-        first: true,
+    /// Returns the length of a combination produced by this iterator.
+    #[inline]
+    pub fn k(&self) -> usize { self.indices.len() }
+
+    /// Returns the (current) length of the pool from which combination elements are
+    /// selected. This value can change between invocations of `next()` and `init()`.
+    #[inline]
+    pub(crate) fn n(&self) -> usize { self.pool.len() }
+
+    /// Returns a reference to the source iterator.
+    #[inline]
+    pub(crate) fn src(&self) -> &I { &self.pool.it }
```

Comment:
Do we really need (all of) these? E.g. at some point, we stored `k` explicitly, and we threw it out because we then always had multiple ways (namely `k` vs. `indices.len` to compute the very same thing), which just bloated the iterator.

Reply:
These are mainly a consequence of handing most of our information over to `Combinations`. We could remove `src()` at the expense of losing a decent `size_hint()` but I think `k()` and `n()` have more justification:

- `k()` and `n()` are used within `Powerset` to detect iterator completion. The alternatives I could see were a `Powerset::done` field (bloat) or just allowing a completed `Powerset` to keep increasing the size of `k` for its `Combinations` on every call to `next()` (wasteful, could allocate).
- A combination's `k` and `n` are also part of the formal notation for combinations in set-theory, so I thought it might make sense to have them visible.
- `k()` and `n()` are also necessary for the `size_hint()` impl.
- Could potentially make code more readable e.g. `self.indices.len()` vs `self.k()` 

I understand that we shouldn't jump through too many hoops for `size_hint` (and there are quite a few hoops here!), but since powersets can become large very quickly as input size increases, having one might be important for users.